web/src/app/components/bar-chart/utils.ts

```diff
@@ -0,0 +1,175 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {formatDate} from '@angular/common';
+import * as d3 from 'd3';
+
+import {step189_2020} from '../../../proto/step189_2020';
+import {DurationItem, findDuration, findDurationUnit, UNIT_CONVERSION} from '../duration-utils';
+
+import {Item} from './bar-chart.component';
+
+/**
+ * D3 types used by the bar chart.
+ *
+ * The d3.Selection has the default type Selection<GElement, Datum, PElement,
+ * PDatum>, and we want to use it with Datum, Datum, PElement, PDatum being
+ * `undefined` or `null`. The SVGSVGElement provides the access and all methods
+ * to manipulate `<svg>` element, while SVGGElement corresponds to the `g`
+ * element that the top bar chart and the bottom bar chart belong to.
+ *
+ * We separate the top bar chart and the bottom bar chart by `g` elements, so
+ * that they can be updated with different methods using dropdown menu and
+ * brush selector.
+ */
+export type d3SVG = d3.Selection<SVGSVGElement, Item[], null, undefined>;
+export type d3G = d3.Selection<SVGGElement, Item[], null, undefined>;
+export type d3Circle =
+    d3.Selection<SVGCircleElement, Item, SVGGElement, Item[]>;
+export type d3HTML = d3.Selection<HTMLDivElement, Item, null, undefined>;
+export type d3Rect = d3.Selection<SVGRectElement, Item, SVGGElement, Item[]>;
+export type d3ScaleLinear = d3.ScaleLinear<number, number>;
+export type d3ScaleBand = d3.ScaleBand<string>;
```

Comment:
If some of these types are not used I would put keep them in the main file. Another option would be to put them in a separate `types.ts` file.